portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -58,21 +76,46 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    // Get name and text
-    String name = request.getParameter("name-input");
-    String text = request.getParameter("text-input");
-    if (text != null && name != null) {
-      long timestamp = System.currentTimeMillis();
-
-      Entity messageEntity = new Entity(Constants.MESSAGE_ENTITY_TYPE);
-      messageEntity.setProperty("name", name);
-      messageEntity.setProperty("text", text);
-      messageEntity.setProperty("timestamp", timestamp);
-
-      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-      datastore.put(messageEntity);
+
+    UserService userService = UserServiceFactory.getUserService();
+    if (userService.isUserLoggedIn()) {
```

Comment:
right now, if a user tries to send a POST request without being logged in, it will succeed silently. How could you fix that? Perhaps throw an error saying "you need to log in oto post a comment"

Reply:
I hide the text prompt for people that are not logged in

Reply:
What if that fails? What if a user decides to call it directly (someone can call an endpoint without using your javascript).

In general, we use an approach called "defence in depth". You treat each part of your code as untrustworthy/buggy, and you write in assertions and helpful error messages to describe when your assumptions about usage or behavior or prerequisites aren't met. This is how you get informative error messages in complex systems - and how we avoid silent bugs. If something isn't supported by your code, don't just silently allow it to be called that way, throw an exception to signal to the caller that they've done something wrong!

Reply:
doPost and doGet are defined with throws IOException, should I be throwing and exception?

Reply:
Yep! I'd recommend throwing an IllegalStateException, or an IllegalArgumentException, or something analgous. Feel free to create your own exceptions to be thrown in these cases!

Reply:
Your error message can have the detail of what assumption was violated.